src/charm.py

```diff
@@ -413,43 +419,55 @@ def _on_hydra_pebble_ready(self, event: WorkloadEvent) -> None:
         """Event Handler for pebble ready event."""
         self._handle_status_update_config(event)
 
+    def _migration_is_needed(self):
+        if not self._peers:
+            return
+
+        return (
+            self._get_peer_data(DB_MIGRATION_VERSION_KEY)
+            != self._hydra_cli.get_version()["version"]
+        )
+
     def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
         """Event Handler for database created event."""
         logger.info("Retrieved database details")
 
-        if not self.unit.is_leader():
-            # TODO: Observe leader_elected event
-            logger.info("Unit does not have leadership")
-            self.unit.status = WaitingStatus("Unit waiting for leadership to run the migration")
-            return
-
         if not self._container.can_connect():
             event.defer()
             logger.info("Cannot connect to Hydra container. Deferring the event.")
             self.unit.status = WaitingStatus("Waiting to connect to Hydra container")
             return
 
+        if not self._peers:
+            self.unit.status = WaitingStatus("Waiting for peer relation")
+            event.defer()
+            return
+
         self.unit.status = MaintenanceStatus(
             "Configuring container and resources for database connection"
         )
 
-        if not self._hydra_service_is_created:
-            event.defer()
-            self.unit.status = WaitingStatus("Waiting for Hydra service")
-            logger.info("Hydra service is absent. Deferring the event.")
-            return
-
         logger.info("Updating Hydra config and restarting service")
+        self._container.add_layer(self._container_name, self._hydra_layer, combine=True)
         self._container.push(self._hydra_config_path, self._render_conf_file(), make_dirs=True)
 
-        try:
-            self._run_sql_migration(set_timeout=True)
-        except ExecError as err:
-            logger.error(f"Exited with code {err.exit_code}. Stderr: {err.stderr}")
+        if not self._migration_is_needed():
+            self._container.start(self._container_name)
+            self.unit.status = ActiveStatus()
+            return
+
+        if not self.unit.is_leader():
+            logger.info("Unit does not have leadership")
+            self.unit.status = WaitingStatus("Unit waiting for leadership to run the migration")
+            event.defer()
```

Comment:
About the status:
This was the behavior of the charm before. I don't think that it makes sense to put the unit in active status until the migration has run.

About the leader elected:
I am not 100% sure what you are suggesting, defer the leader elected event until the migration has run and run the migration from that hook? I think that we need to defer this event in case something happens to the leader, eg:
you have 2 units (A, B) and unit A is leader. The database is created and the `_on_database_created` hook runs. Let's assume the the hook is fired in unit B before unit A. Let's also assume that for some reason unit A dies before running the migration. Another unit is spin up, but unit B is elected as the leader. IF unit B did not defer the event, the migration will never run.

Reply:
What I meant was to put the non-leader units in active state if the leader already run the migration, but that's a mistake on my side, the previous check `_migration_is_needed` would change the status if migration was already executed, so please dismiss my comment.
The second comment also makes sense. We could run the migration on leader elected, but then we'd need a way to check if it has already run by some other unit.